src/impl_x86__base_implementation.inl

```diff
@@ -13,6 +13,65 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+// A note on x86 SIMD instructions availability
+// -----------------------------------------------------------------------------
+// A number of conditions need to be met for an application to use SIMD
+// instructions:
+// 1. The CPU itself must support the instruction.
+// - we use `CPUID` to check whether the feature is supported.
+// 2. The OS must save and restore the associated SIMD register across context
+//    switches, we check that:
+// - the CPU reports supporting hardware context switching instructions via
+//   CPUID.1:ECX.XSAVE[bit 26]
+// - the OS reports supporting hardware context switching instructions via
+//   CPUID.1:ECX.OSXSAVE[bit 27]
+// - the CPU extended control register 0 (XCR0) is set to save and restore the
+//   needed SIMD registers
+//
+// Note that if `XSAVE`/`OSXSAVE` are missing, we delegate the detection to the
+// OS via the `DetectFeaturesFromOs` function or via microarchitecture
+// heuristics.
+//
+// Encoding
+// -----------------------------------------------------------------------------
+// `X86Info` contains fields such as `vendor` and `brand_string` they are
+// represents characters in ASCII encoding. `vendor` length of characters is 13
+// and `brand_string` is 49 (with null terminated string). We use
+// CPUID.1:E[D,C,B]X to get `vendor` and CPUID.8000_000[4:2]:E[D,C,B,A]X to get
+// `brand_string`
+//
+// Microarchitecture
+// -----------------------------------------------------------------------------
+// `GetX86Microarchitecture` function consists of check on vendor via
+// `IsVendorByX86Info`. We use `CPUID(family, model)` macro to define
+//  microarchitecture of vendor. In cases where the `family` and `model` is the
+//  same for several microarchitectures we do a stepping check or in the worst
+//  case parsing `brand_string` as example `HasSecondFMA`. Details of
+//  identification by `brand_string` can be found in the `Identification`
+//  section by reference:
+//  https://en.wikichip.org/wiki/intel/microarchitectures/cascade_lake
+//
+// CacheInfo X86
+// -----------------------------------------------------------------------------
+// We use common struct `CacheInfo` for all CPU architectures and this struct
```

Comment:
"We use the `CacheInfo` struct to store information about cache levels. The maximum number of levels is hardcoded but can be increased if needed. We have full support .,."